Step3: 자동차 경주

programmer-sjk opened the pull request:
안녕하세요 선호님 Step 3 리뷰 요청 드립니다.

한 가지 의문이 있는데요 아래 요구사항은 둘 중 무엇을 의미하는지 궁금합니다.
**모든 로직에 단위 테스트를 구현한다.**
1. 모든 함수를 public으로 열고 테스트 코드를 추가한다.
2. private 함수를 사용하되, 모든 public 함수는 테스트 코드를 추가한다.

우선은 2번이라 생각하고 코드를 구현하였습니다.
테스트 코드가 작아, 구조를 잘못 만들었나? 라는 생각도 들었는데 선호님의 의견 부탁드립니다.

programmer-sjk committed:
Step3: 자동차 경주

- Racing, RacingCar, Show 클래스 구현
- 기능 구현

programmer-sjk committed:
Step3: 자동차 경주

- 변수 네이밍 변경
- 매직 넘버 상수화

programmer-sjk committed:
Step3: 자동차 경주

- 변수 네이밍 변경
- 매직 넘버 상수화

programmer-sjk committed:
Step3: 자동차 경주

- 테스트 코드 추가
- lint 적용

vsh123 commented on src/main/kotlin/racing/RacingView.kt:
```diff
@@ -0,0 +1,26 @@
+object RacingView {
+    private val DISPLAY_NUMBER = 1
+    fun result(round: Int, carDistances: ArrayList<Array<Int>>) {
+        for(round in 0 until round) {
```
* 린트가 적용되어있지 않은 것 같아요! 린트확인 한번 부탁드릴게요!
* round라는 네이밍이 함수의 인자 `round`와 중복되는 것 같아요! 네이밍을 변경해보면 어떨까요?

programmer-sjk replied:
앗..! 돌렸다고 생각했는데 다시 돌려보니 에러가 있네요 ㅜㅜ 죄송합니다

vsh123 commented on src/main/kotlin/racing/RacingView.kt:
```diff
@@ -0,0 +1,26 @@
+package racing
+
+object RacingView {
+    private val DISPLAY_NUMBER = 1
```
[const](https://kimch3617.tistory.com/entry/Kotlin%EC%97%90%EC%84%9C-val-%EC%99%80-const-val-%EC%9D%98-%EC%B0%A8%EC%9D%B4)에 대해 알아보고 적용해보면 어떨까요?

programmer-sjk replied:
넵 반영했습니다.

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,25 @@
+package racing
```
step2의 작업 커밋이 반영되어 있지 않은 것 같아요!
rebase를 통해 커밋최신화를 해보면 어떨까요?

https://nextstep-members.slack.com/archives/C03CB32ALKZ/p1652064793557939

programmer-sjk replied:
넵 rebase 했습니다.

vsh123 commented on src/main/kotlin/racing/main.kt:
```diff
@@ -0,0 +1,10 @@
+
+fun main() {
+    println("자동차 대수는 몇 대인가요?")
+    val carNumber = readLine()!!.toInt()
```
* `!!` not null 키워드는 최대한 지양하는게 어떨까요??  null이 들어온다면 NullPointException이 발생할 수 있거든요! [readln](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/readln.html)을 활용해보세요!
* 숫자가 아닌 다른 값이 들어왔을 때에 데한 에러 핸들링을 하면 어떨까요?

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,25 @@
+
+object Racing {
+    fun start(carNumber: Int, tryRound: Int): ArrayList<Array<Int>> {
+        val racingCars = Array<RacingCar>(carNumber) { RacingCar(tryRound) }
```
Array보다는 Collection을 활용해보면 어떨까요? 어떤 이점이 있을까요?

```suggestion
        val racingCars = List(carNumber) { RacingCar(tryRound) }
```

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,25 @@
+package racing
+
+object Racing {
+    fun start(carNumber: Int, tryRound: Int): ArrayList<Array<Int>> {
```
리턴값은 구현체(ArrayList)가 아닌 인터페이스를 리턴해주면 어떨까요?

```suggestion
    fun start(carNumber: Int, tryRound: Int): List<Array<Int>> {
```

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,25 @@
+    fun start(carNumber: Int, tryRound: Int): ArrayList<Array<Int>> {
+        val racingCars = Array<RacingCar>(carNumber) { RacingCar(tryRound) }
+
+        for (count in 0 until tryRound) {
```
[repeat](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/repeat.html)를 활용해보면 어떨까요?

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -0,0 +1,29 @@
+
+class RacingCar(private val round: Int) {
+    private val distances = Array(round) {0}
+    private val MIN_PIVOT_VALUE = 0
```
class의 상수는 top-level이나 companion object에 선언해주면 어떨까요?
[참고 ](https://stackoverflow.com/questions/44038721/constants-in-kotlin-whats-a-recommended-way-to-create-them)

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -0,0 +1,29 @@
+import kotlin.random.Random
+
+class RacingCar(private val round: Int) {
+    private val distances = Array(round) {0}
```
여기도 Array보다는 Collectinon을 활용해보면 어떨까요?

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -0,0 +1,29 @@
+        return distances
+    }
+    fun race() {
+        if (canMove()) {
```
`canMove`라는 랜덤 함수가 들어감에 따라 해당 메서드에 대한 테스트가 어려워진 것 같아요!
[이 글](https://www.slipp.net/questions/557)을 읽어보고 테스트가능한 구조로 리팩토링에 도전해보세요!

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -0,0 +1,29 @@
+        if (canMove()) {
+            move()
+        }
+        currentRound += 1
```
1이라는 값은 상수화를 안하신 이유가 있을까요?

programmer-sjk replied:
모든 literal은 상수화가 되어야 하는군요..!
현재 라운드에 1을 더한다는 개념이 쉬워 보여서 나뒀는데 
상수화해서 리뷰 요청 드리겠습니다.

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -0,0 +1,29 @@
+    private val MIN_PIVOT_VALUE = 0
+    private val MAX_PIVOT_VALUE = 10
+    private val MOVE_PIVOT = 4
+    var currentRound = 0
```
`var`인 currentRound를 외부에서 set하지 못하게 막아보면 어떨까요?
[참고](https://codechacha.com/ko/kotlin-property/#private-setter-%EB%A7%8C%EB%93%A4%EA%B8%B0)

vsh123 reviewed (changes requested):
안녕하세요!
step3 구현 잘해주셨네요~!!

몇가지 코멘트를 드렸으니 확인부탁드려요!!
또한 아래와 같은 부분을 집중적으로 확인 부탁드려요!

* 요청전에 lint확인 부탁드려요~!
* Array로 선언되어 있는 부분을 Collection전환으로 도전해보세요!
* ArrayList와 같이 구현체로 리턴하거나 받는 부분을 List 인터페이스로 변경해보세요!

programmer-sjk committed:
Step3: 자동차 경주 리뷰 반영

- const 적용
- Array 대신 Coolection 활용
- repeat 사용
- 매직넘버 상수화
- class 상수에 companion object 사용
- lint 적용

programmer-sjk committed:
Step3: 자동차 경주 리뷰 반영

- 리뷰 반영으로 인한 테스트 코드 추가

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -0,0 +1,31 @@
+        const val NEXT_ROUND = 1
+    }
+
+    private val distances = MutableList(round) {0}
```
여기로 린트 적용한번 부탁드려요~

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -0,0 +1,31 @@
+    private val distances = MutableList(round) {0}
+    private var currentRound = 0
+
+    fun getDistances(): MutableList<Int> {
```
getter를 별도로 만들 필요가 있을까요? [properties getter](https://kotlinlang.org/docs/properties.html#getters-and-setters)를 활용해보세요!

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -0,0 +1,34 @@
+
+import kotlin.random.Random
+
+object Racing {
```
해당 클래스에 있는 MutableList / ArrayList 를 모두 List로 변경에 도전해보세요~!!

vsh123 commented on src/test/kotlin/racing/RacingTest.kt:
```diff
@@ -0,0 +1,24 @@
+
+class RacingTest {
+    @Test
+    fun `race를 시작하면 반환되는 2차 배열의 길이는 자동차 개수와 같다`() {
```
더이상 리턴값이 배열이 아닌 것 같아요!

vsh123 reviewed (approved):
안녕하세요~!!!
피드백 반영 잘해주셨네요!! :)

몇가지 코멘트를 남겼으니 다음스텝에서 함께 반영해보세요!
이번 스텝은 머지하도록 하겠습니다~!
